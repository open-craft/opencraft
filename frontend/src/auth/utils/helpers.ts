import { V2Api } from 'global/api';

export const checkAuthAndRefreshToken = async () => {
  const refreshToken = localStorage.getItem('token_refresh');
  let response = null;

  if (refreshToken) {
    try {
      response = await V2Api.authRefreshCreate({
        data: { refresh: refreshToken }
      });
    } catch (e) {
      return false;
    }
  }

  if (response) {
    localStorage.setItem('token_access', response.access);
    return true;
  }

  return false;
};
