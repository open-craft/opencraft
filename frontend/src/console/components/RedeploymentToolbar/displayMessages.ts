const messages = {
  deploy: {
    defaultMessage: 'Deploy ({undeployedChanges} updates)',
    description: 'Deployment button text.'
  },
  cancelDeploymentConfirm: {
    defaultMessage: 'Are you sure you want to cancel this redeployment?',
    description: ''
  },
  cancelDeploymentConfirmText: {
    defaultMessage:
      'Your instance is being updated with the latest settings. If you cancel ' +
      'this deployment, your changes wont be lost, but they will need to be ' +
      'redeployed.',
    description: ''
  },
  closeModalButton: {
    defaultMessage: 'Close',
    description: ''
  },
  confirmCancelDeployment: {
    defaultMessage: 'Cancel redeployment',
    description: ''
  }
};

export default messages;
