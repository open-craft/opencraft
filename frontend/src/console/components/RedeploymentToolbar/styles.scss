@import '~styles/theme';


.redeployment-toolbar {
  background-color: $light-blue;
  padding: 15px;
  margin: 0 -15px;

  display: flex;
  align-items: center;

  .redeployment-nav {
    display: inline-block;
    width: 100%;
    max-width: 1100px;

    .status-pill {
      background-color: #ddd;
      padding: 8px 20px;
      margin: 4px 2px;
      border-radius: 39px;
      text-align: center;
      display: inline-block;
      cursor: pointer;
      background-color: $primary-2;
      height: 44px;

      .text {
        font-family: $roboto-font;
        font-weight: 500;
        font-size: 18px;
        color: #ffffff;
      }

      .dot {
        height: 25px;
        width: 25px;
        border: 2px solid black;
        background-color: lightgreen;
        border-radius: 50%;
        display: inline-block;
      }
    }
  }
}
