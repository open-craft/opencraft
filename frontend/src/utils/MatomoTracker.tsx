import React, { ReactNode } from 'react';
import {
  createInstance,
  MatomoProvider,
  useMatomo
} from '@datapunt/matomo-tracker-react';
import {
  MATOMO_ALIAS_DOMAIN,
  MATOMO_BASE_URL,
  MATOMO_MY_DOMAIN,
  MATOMO_SITE_ID
} from '../global/constants';

interface MatomoTrackerProps {
  children: ReactNode;
}

interface MatomoAdditionalConfiguration {
  [propName: string]: any;
}

interface MatomoConfiguration {
  urlBase: string;
  siteId: number;
  configurations?: MatomoAdditionalConfiguration;
}

interface MatomoUserIdTrackerProps {
  userId: string;
}

/*
 * This wrapper component initializes the Matomo tracking by wrapping the children
 * elements, if the tracking is enabled by providing a valid value for the 'MATOMO_BASE_URL'
 * constant. It allows tracking page views, events and the user id, when available.
 *
 * The 'MATOMO_ALIAS_DOMAIN' constant allows the cross-domain linking of another site with
 * this site so that they are treated as the same site in Matomo.
 */
export const MatomoTracker: React.FC<MatomoTrackerProps> = (
  props: MatomoTrackerProps
) => {
  let element = props.children;
  if (MATOMO_BASE_URL) {
    const matomoConfiguration: MatomoConfiguration = {
      urlBase: MATOMO_BASE_URL,
      siteId: parseInt(MATOMO_SITE_ID, 10)
    };
    if (MATOMO_ALIAS_DOMAIN) {
      const myDomain = `*.${MATOMO_MY_DOMAIN}`;
      const aliasDomain = `*.${MATOMO_ALIAS_DOMAIN}`;
      matomoConfiguration.configurations = {
        enableCrossDomainLinking: true,
        setDomains: [myDomain, aliasDomain]
      };
    }
    element = (
      <MatomoProvider value={createInstance(matomoConfiguration)}>
        {props.children}
      </MatomoProvider>
    );
  }
  return <>{element}</>;
};

/**
 * This is a component passes the provided user id to Matomo if Matomo tracking is enabled.
 * The 'pushInstruction' function exposed by the 'useMatomo()' hook is used to do this.
 *
 * This component is needed as all the components in the registration process are classes and hence
 * the 'yseMatomo' hook cannot be directly used in those components.
 */
export const MatomoUserIdTracker: React.FC<MatomoUserIdTrackerProps> = (
  props: MatomoUserIdTrackerProps
) => {
  const { pushInstruction } = useMatomo();
  if (MATOMO_BASE_URL && props.userId) {
    pushInstruction('setUserId', props.userId);
  }
  return null;
};
