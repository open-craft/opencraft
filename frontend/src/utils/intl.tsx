import React from "react";
import { FormattedMessage, MessageDescriptor } from "react-intl";

type MessageMap = { [k: string]: Omit<MessageDescriptor, 'id'> };

interface Props<T, K extends keyof T> {
    id: K;
    messages: T;
    values?: any;
}


export const WrappedMessage = (props: Props<MessageMap, string>) => {
    const {messages, id, ...values} = props;
    return (
        <FormattedMessage  {...messages[id]} id={id} {...values}/>
    );
};
