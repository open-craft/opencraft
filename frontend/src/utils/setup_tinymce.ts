import 'tinymce/tinymce';
import 'tinymce/themes/silver';
import 'tinymce/skins/ui/oxide/skin.min.css';
import 'tinymce/skins/ui/oxide/content.min.css';
import 'tinymce/plugins/advlist';
import 'tinymce/plugins/autolink';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/link';
import 'tinymce/plugins/image';
import 'tinymce/plugins/charmap';
import 'tinymce/plugins/print';
import 'tinymce/plugins/preview';
import 'tinymce/plugins/anchor';
import 'tinymce/plugins/searchreplace';
import 'tinymce/plugins/visualblocks';
import 'tinymce/plugins/code';
import 'tinymce/plugins/fullscreen';
import 'tinymce/plugins/insertdatetime';
import 'tinymce/plugins/media';
import 'tinymce/plugins/table';
import 'tinymce/plugins/paste';
import 'tinymce/plugins/help';
import 'tinymce/plugins/wordcount';
