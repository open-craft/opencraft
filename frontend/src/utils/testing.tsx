import { createMemoryHistory } from 'history';
import * as React from 'react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import * as renderer from 'react-test-renderer';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { createRootReducer } from '../global/reducers';

export const setupComponentForTesting = (
  reactContent: JSX.Element,
  storeContents = {}
) => {
  const middleware = applyMiddleware(thunk);

  const store = createStore(
    createRootReducer(createMemoryHistory()),
    storeContents,
    middleware
  );

  return renderer.create(
    <IntlProvider textComponent={React.Fragment} locale="en">
      <Provider store={store}>
        <MemoryRouter>{reactContent}</MemoryRouter>
      </Provider>
    </IntlProvider>
  );
};
