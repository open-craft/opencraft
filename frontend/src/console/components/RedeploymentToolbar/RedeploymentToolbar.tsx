import * as React from 'react';
import { WrappedMessage } from 'utils/intl';
import { Button, Modal } from 'react-bootstrap';
import { CustomStatusPill } from 'ui/components';
import { DeploymentInfoModel } from 'console/models';
import {
  OpenEdXInstanceDeploymentStatusStatusEnum as DeploymentStatus,
  OpenEdXInstanceDeploymentStatusDeploymentTypeEnum as DeploymentType
} from 'ocim-client';
import messages from './displayMessages';
import './styles.scss';

interface Props {
  deployment?: DeploymentInfoModel;
  cancelRedeployment: Function;
  performDeployment: Function;
  loading: boolean;
}

export const RedeploymentToolbar: React.FC<Props> = ({
  deployment,
  cancelRedeployment,
  performDeployment,
  loading
}: Props) => {
  const [show, setShow] = React.useState(false);

  const handleCloseModal = () => setShow(false);
  const handleShowModal = () => setShow(true);

  let deploymentDisabled: boolean = true;
  let cancelDeploymentDisabled: boolean = true;
  let undeployedChanges: number = 0;
  let deploymentStatus: DeploymentStatus | null = null;
  let deploymentType: DeploymentType | null = null;

  if (deployment) {
    deploymentStatus = deployment.status;
    undeployedChanges = deployment.undeployedChanges.length;
    deploymentType = deployment.deploymentType;
    /**
     * The user can't trigger a deployment when:
     * 1. There's a pending request.
     * 2. The instance is being prepared.
     * 3. There are no new changes to deploy.
     * 4. There's already a provisioning running.
     */
    deploymentDisabled =
      loading ||
      !undeployedChanges ||
      deploymentStatus === DeploymentStatus.Preparing;

    /**
     * The user can't cancel a deployment when:
     * 1. There's a pending request.
     * 2. There's no provisioning running.
     * 3. There's a provisionin running, but it wasn't triggered
     *    by the user.
     */
    cancelDeploymentDisabled =
      loading ||
      deploymentStatus !== DeploymentStatus.Provisioning ||
      (deploymentStatus === DeploymentStatus.Provisioning &&
        deploymentType !== DeploymentType.User);
  }

  const cancelDeploymentHandler = cancelDeploymentDisabled
    ? undefined
    : handleShowModal;

  return (
    <div className="d-flex justify-content-center align-middle redeployment-toolbar">
      <div className="redeployment-nav">
        <CustomStatusPill
          loading={loading}
          redeploymentStatus={deploymentStatus}
          deploymentType={deploymentType}
          cancelRedeployment={cancelDeploymentHandler}
        />

        <Button
          className="float-right loading"
          variant="primary"
          size="lg"
          onClick={() => {
            performDeployment();
          }}
          disabled={deploymentDisabled}
        >
          <WrappedMessage
            id="deploy"
            messages={messages}
            values={{ undeployedChanges }}
          />
        </Button>
      </div>

      <Modal
        show={show}
        onHide={handleCloseModal}
        className="cancel-redeployment-modal"
        centered
      >
        <Modal.Header>
          <Modal.Title>
            <p>
              <WrappedMessage
                id="cancelDeploymentConfirm"
                messages={messages}
              />
            </p>
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <WrappedMessage
            id="cancelDeploymentConfirmText"
            messages={messages}
          />
        </Modal.Body>
        <Modal.Footer>
          <Button
            variant="outline-primary"
            size="lg"
            onClick={handleCloseModal}
          >
            <WrappedMessage id="closeModalButton" messages={messages} />
          </Button>
          <Button
            variant="primary"
            size="lg"
            onClick={() => {
              cancelRedeployment();
              handleCloseModal();
            }}
          >
            <WrappedMessage id="confirmCancelDeployment" messages={messages} />
          </Button>
        </Modal.Footer>
      </Modal>
    </div>
  );
};
