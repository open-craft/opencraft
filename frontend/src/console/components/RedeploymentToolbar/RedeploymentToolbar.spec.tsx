import React from 'react';
import { setupComponentForTesting } from "utils/testing";
import { RedeploymentToolbar } from './RedeploymentToolbar';

it('renders without crashing', () => {
    const tree = setupComponentForTesting(<RedeploymentToolbar />).toJSON();
    expect(tree).toMatchSnapshot();
});
