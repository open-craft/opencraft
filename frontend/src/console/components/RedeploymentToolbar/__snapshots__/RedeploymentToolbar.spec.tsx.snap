// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedeploymentToolbar Component Correctly renders redeployment bar when deployment info isn't there 1`] = `
<div
  className="d-flex justify-content-center align-middle redeployment-toolbar"
>
  <div
    className="redeployment-nav"
  >
    <span
      className="status-pill badge badge-pill badge-primary"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        className="dot"
        style={
          Object {
            "backgroundColor": "grey",
          }
        }
      />
      <div
        className="text"
      >
        Status: loading
      </div>
    </span>
    <button
      className="float-right loading btn btn-primary btn-lg"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Deploy (0 updates)
    </button>
  </div>
</div>
`;

exports[`RedeploymentToolbar Component Correctly renders redeployment bar when instance has pending changes 1`] = `
<div
  className="d-flex justify-content-center align-middle redeployment-toolbar"
>
  <div
    className="redeployment-nav"
  >
    <span
      className="status-pill badge badge-pill badge-primary"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        className="dot"
        style={
          Object {
            "backgroundColor": "#1abb64",
          }
        }
      />
      <div
        className="text"
      >
        Status: Pending changes
      </div>
    </span>
    <button
      className="float-right loading btn btn-primary btn-lg"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Deploy (3 updates)
    </button>
  </div>
</div>
`;

exports[`RedeploymentToolbar Component Correctly renders redeployment bar when instance is being deployed 1`] = `
<div
  className="d-flex justify-content-center align-middle redeployment-toolbar"
>
  <div
    className="redeployment-nav"
  >
    <span
      className="status-pill badge badge-pill badge-primary"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        className="dot"
        style={
          Object {
            "backgroundColor": "#ff9b04",
          }
        }
      />
      <div
        className="text"
      >
        Status: deploying changes
      </div>
      <div
        className="text cancel-deployment nav"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <i
          className="fas fa-xs fa-times"
        />
        Cancel deployment
      </div>
    </span>
    <button
      className="float-right loading btn btn-primary btn-lg"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Deploy (0 updates)
    </button>
  </div>
</div>
`;

exports[`RedeploymentToolbar Component Correctly renders redeployment bar when instance is being prepared 1`] = `
<div
  className="d-flex justify-content-center align-middle redeployment-toolbar"
>
  <div
    className="redeployment-nav"
  >
    <span
      className="status-pill badge badge-pill badge-primary"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        className="dot"
        style={
          Object {
            "backgroundColor": "#ff9b04",
          }
        }
      />
      <div
        className="text"
      >
        Status: We're preparing your instance 
      </div>
    </span>
    <button
      className="float-right loading btn btn-primary btn-lg"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Deploy (0 updates)
    </button>
  </div>
</div>
`;

exports[`RedeploymentToolbar Component Correctly renders redeployment bar when instance is up-to-date 1`] = `
<div
  className="d-flex justify-content-center align-middle redeployment-toolbar"
>
  <div
    className="redeployment-nav"
  >
    <span
      className="status-pill badge badge-pill badge-primary"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <span
        className="dot"
        style={
          Object {
            "backgroundColor": "#1abb64",
          }
        }
      />
      <div
        className="text"
      >
        Status: Up to date
      </div>
    </span>
    <button
      className="float-right loading btn btn-primary btn-lg"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Deploy (0 updates)
    </button>
  </div>
</div>
`;
